src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -229,6 +232,33 @@ def safe_to_destroy(self, ids):
                 'is_safe_to_destroy': False,
             }
 
+    @Endpoint('GET', query_params=['ids'])
+    @ReadPermission
+    def safe_to_remove(self, ids):
+        """
+        :type ids: int|[int]
+        """
+        # Unused now, might be useful for individual OSD checks in the future
+        _ = ids
+
+        # Check if there are any OSDs exceeds nearfull or full threshold.
+        all_health = health.HealthData(self._has_permissions, minimal=True).all_health()
+        checks = {c['type'] for c in all_health['health']['checks']}
+        unsafe_checks = set(['OSD_FULL', 'OSD_BACKFILLFULL', 'OSD_NEARFULL'])
+        failed_checks = checks & unsafe_checks
+        return {
+            'is_safe_to_remove': not bool(failed_checks),
+            'message': 'Removing OSD(s) is not recommended with these health checks: {}.'.
+                       format(', '.join(failed_checks)) if failed_checks else ''
+        }
+
+    @Endpoint('POST')
+    @DeletePermission
+    @handle_orchestrator_error('osd')
+    def remove_osds(self, osd_ids):
```

Comment:
As [already mentioned](https://github.com/ceph/ceph/pull/31997/files#r353945856), this should be a `DELETE /api/osd/<osd_id>`. I don't think we should support multi-resource deletion (not HTTP 1.1 compliant): OSDs would be mostly deleted 1 at a time (or all OSDs in a host/rack, ~100s?), and, in worst, max number of OSDs won't exceed the range of thousands (~1000's HTTP request to delete all cluster OSDs I think is bearable, as we are generating a similar amount of background requests in 10-20 minutes with no user interaction).

Reply:
Fixed.
- Rewording `Remove` to `Delete` in the frontend for consistency. We have `Create/Delete` pools/images in Dashboard. Let's use the same verbs.
- Now it only supports removing single OSD at one time.
  - `DELETE /api/osd/<svc_id>`: delete `osd.svc_id`. An error is returned if
there are any pre-checks fail.
  - `DELETE /api/osd/<svc_id>?force=true`: with `force` flag on, the
deleting request is sent to orchestrator even pre-checks fail.
 - I'm not sure if it's appropriate to DELETE a resource with query parameters.

Reply:
`DELETE` can be invoked with a query string, but just googled it and it seems this specific use of it [is pretty debated](https://stackoverflow.com/questions/2539394/rest-http-delete-and-parameters).

Reply:
> As [already mentioned](https://github.com/ceph/ceph/pull/31997/files#r353945856), this should be a `DELETE /api/osd/<osd_id>`. I don't think we should support multi-resource deletion (not HTTP 1.1 compliant): OSDs would be mostly deleted 1 at a time (or all OSDs in a host/rack, ~100s?), and, in worst, max number of OSDs won't exceed the range of thousands (~1000's HTTP request to delete all cluster OSDs I think is bearable, as we are generating a similar amount of background requests in 10-20 minutes with no user interaction).

I need to start with saying that I don't know much about RESTful designs and HTTP 1.1 compliance and this is not about trying to convince you to change anything, but it seems that there are several opinions ([[1]](https://stackoverflow.com/questions/21863326/delete-multiple-records-using-rest) [[2]](https://stackoverflow.com/questions/511281/patterns-for-handling-batch-operations-in-rest-web-services?answertab=votes#tab-top)) out there about how to do RESTful bulk operations. POST seems to be an option. I wonder why we want to allow the front-end to make hundreds of requests in an admittedly rare situation if we could do it in just a single request. So I think my questions are:

Is it really not possible to do it compliant with HTTP 1.1 and secondly, why would HTTP 1.1 compliance be so important?

Reply:
@p-se, answering your specific questions:

1. Yes, it is possible. In fact, I'm almost sure we're not 100% HTTP 1.1 compliant.

2. Because it's the most used standard in the once-called WWW, and if we want to claim that we have an open and stable API, we better ensure that the API complies with some standard.

Going a bit deeper into the bulk operations debate, I can see a few reasons to keep our (90%-REST) API without them:

- First and foremost, **simplicity**: single-op requests are easier to implement, test, tune, log, trace and debug.
- Bulk ops will likely turn into Async ops. Providing Bulk Async Ops on a non-transaction-safe/non-persistent back-end may work perfectly when the sun is shining, but it's less predictable under non ideal (field) conditions.
- And, we loose the benefits from leveraging a standardized behaviour: interoperability, design & user docs, code/tool reuse...